File: python/levenshtein.py
#!/usr/bin/env python
# -*- coding: cp932 -*-
def levenshtein_distance(a , b):
    """
        >>> import levenshtein
        >>> levenshtein.levenshtein_distance('bcd','efg')
        3
        >>> levenshtein.levenshtein_distance('abc', 'abc')
        0
        >>> levenshtein.levenshtein_distance('kitten', 'sitting')
        3
        >>> levenshtein.levenshtein_distance('smei', 'mei')
        1
        >>> levenshtein.levenshtein_distance('12345', '234')
        2
        >>> levenshtein.levenshtein_distance('おしり', 'めがね')
        3
        >>> levenshtein.levenshtein_distance('照明', '照明')
        0
        >>> levenshtein.levenshtein_distance('いじめ', 'いじり')
        1
        >>> levenshtein.levenshtein_distance('もずくのかに', 'かに')
        4
        >>> levenshtein.levenshtein_distance('CDショップ', 'Cショック')
        2
        >>> levenshtein.levenshtein_distance('火事', '花火')
	2
    """

    m = [ [0] * (len(b) + 1) for i in range(len(a) + 1) ]

    for i in xrange(len(a) + 1):
        m[i][0] = i

    for j in xrange(len(b) + 1):
        m[0][j] = j

    for i in xrange(1, len(a) + 1):
        for j in xrange(1, len(b) + 1):
            if a[i - 1] == b[j - 1]:
                x = 0
            else:
                x = 1
            m[i][j] = min(m[i - 1][j] + 1, m[i][ j - 1] + 1, m[i - 1][j - 1] + x)
    # print m
    return m[-1][-1]

#import sys

#s1 = sys.argv[1]
#s2 = sys.argv[2]

#print levenshtein_distance(s1, s2)
#print levenshtein_distance(unicode(s1, sys.stdin.encoding), unicode(s2, sys.stdin.encoding))
def _test():
    import doctest
    doctest.testmod()

if __name__ == "__main__":
    _test()

File: lib/levenshtein.php
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>EC Dragon Test</title>
</head>
<body>
<pre>
<?php
/**
 * マルチバイト文字列、英数字の混じった文字列を１文字ずつ配列に分割
 * 参考:http://detail.chiebukuro.yahoo.co.jp/qa/question_detail/q1417635014#
 * mb_split, str_split, preg_splitことごとく使えない。
 */
function mbStringToArray($string, $encoding = 'UTF-8') {
	$arrayResult = array();
	while ($iLen = mb_strlen($string, $encoding)) {
		array_push($arrayResult, mb_substr($string, 0, 1, $encoding));
		$string = mb_substr($string, 1, $iLen, $encoding);
	}
	return $arrayResult;
}

/**
 * 編集距離（レーベンシュタイン距離）を求める（マルチバイト文字対応）
 * @param $str1
 * @param $str2
 * @param $encoding
 * @param $costReplace
 * @return 数値（距離）,かぶっていた文字の数
 */
function LevenshteinDistance($str1, $str2, $costReplace = 2, $encoding = 'UTF-8') {
	$count_same_letter = 0;
	$d = array();
	$mb_len1 = mb_strlen($str1, $encoding);
	$mb_len2 = mb_strlen($str2, $encoding);

	$mb_str1 = mbStringToArray($str1, $encoding);
	$mb_str2 = mbStringToArray($str2, $encoding);

	for ($i1 = 0; $i1 <= $mb_len1; $i1++) {
		$d[$i1] = array();
		$d[$i1][0] = $i1;
	}

	for ($i2 = 0; $i2 <= $mb_len2; $i2++) {
		$d[0][$i2] = $i2;
	}

	for ($i1 = 1; $i1 <= $mb_len1; $i1++) {
		for ($i2 = 1; $i2 <= $mb_len2; $i2++) {
			//			$cost = ($str1[$i1 - 1] == $str2[$i2 - 1]) ? 0 : 1;
			if ($mb_str1[$i1 - 1] === $mb_str2[$i2 - 1]) {
				$cost = 0;
				$count_same_letter++;
			} else {
				$cost = $costReplace; //置換
				}
			$d[$i1][$i2] = min($d[$i1 - 1][$i2] + 1, //挿入
			$d[$i1][$i2 - 1] + 1, //削除
			$d[$i1 - 1][$i2 - 1] + $cost);
		}
	}
	//return $d[$mb_len1][$mb_len2];
	return array('distance' => $d[$mb_len1][$mb_len2], 'count_same_letter' => $count_same_letter);
}

$test = array(
	array('abc', 'abc'), //0
	array('kitten', 'sitting'), //3
	array('おしり', 'めがね'), //5
	array('照明', '明'), //
	array('いじめ', 'いじり'), //
	array('もずくのかに', 'かに'), //
	array('火事', '花火'), //
	array('火事', '黒板'), //
	array('smei', 'mei'), //
	array('12345', '234'),
);
foreach ($test as $row) {
	echo $row[0].','.$row[1];
	echo ' | ';
	echo levenshtein($row[0], $row[1]);
	echo ' = ';
	var_dump(LevenshteinDistance($row[0], $row[1]));
	echo '<br/>';
}

var_dump($test);
?>

</pre>
</body>
</html>